src/Components/WebView/WebView/src/WebViewManager.cs

```diff
@@ -0,0 +1,184 @@
+            // add it when the page attaches later.
+            return _currentPageContext != null
+                ? _currentPageContext.Renderer.AddRootComponentAsync(componentType, selector, parameters)
+                : Task.CompletedTask;
+        }
+
+        /// <summary>
+        /// Removes a previously-attached root component from the current page.
+        /// </summary>
+        /// <param name="selector">The CSS selector describing where in the page the component was placed. This must exactly match the selector provided on an earlier call to <see cref="AddRootComponentAsync(Type, string, ParameterView)"/>.</param>
+        public Task RemoveRootComponentAsync(string selector)
+        {
+            if (!_rootComponentsBySelector.Remove(selector))
+            {
+                throw new InvalidOperationException($"There is no root component with selector '{selector}'.");
+            }
+
+            // If the page is already attached, remove the root component from it now. Otherwise it's
+            // enough to have updated the dictionary.
+            return _currentPageContext != null
+                ? _currentPageContext.Renderer.RemoveRootComponentAsync(selector)
+                : Task.CompletedTask;
+        }
+
+        /// <summary>
+        /// Notifies the <see cref="WebViewManager"/> about a message from JavaScript running within the web view.
+        /// </summary>
+        /// <param name="message">The message.</param>
+        protected void MessageReceived(Uri sourceUri, string message)
+        {
+            if (!_appBaseUri.IsBaseOf(sourceUri))
+            {
+                // It's important that we ignore messages from other origins, otherwise if the webview
+                // navigates to a remote location, it could send commands that execute locally
+                return;
+            }
+
+            _ = _dispatcher.InvokeAsync(async () =>
+            {
+                // TODO: Verify this produces the correct exception-surfacing behaviors.
+                // For example, JS interop exceptions should flow back into JS, whereas
+                // renderer exceptions should be fatal.
+                try
+                {
+                    await _ipcReceiver.OnMessageReceivedAsync(_currentPageContext, message);
+                }
+                catch (Exception ex)
+                {
+                    _ipcSender.NotifyUnhandledException(ex);
+                    throw;
+                }
+            });
+        }
+
+        protected bool TryGetResponseContent(string uri, bool isPageLoad, out int statusCode, out string statusMessage, out Stream content, out string headers)
+            => _staticContentProvider.TryGetResponseContent(uri, isPageLoad, out statusCode, out statusMessage, out content, out headers);
+
+        internal async Task AttachToPageAsync(string baseUrl, string startUrl)
```

Comment:
Also, "spicy" thought is that we need to have a "root component" SetParameters API here to dispatch parameter updates to top-level components.

Reply:
Yes, if this becomes a common scenario then I definitely agree. However the usability of mutating root component parameters from a WPF/WinForms control might not be great, and there will always be alternatives like sending commands via some DI service. I'd suggest we wait to see some specific customer demand here before implementing this.